File: Lab/Day-3/Queue/C++/CircularQueue.cpp
// Circular QUEUE
// Queue as ADT in C++.

#include<iostream>
#include<stdexcept>
#include<array>
#include<utility>

template <typename T, size_t Capacity>

class CircularQueue{
private:
  std::array<T, Capacity> m_data{};
  size_t m_front{0};
  size_t m_rear{0};
  size_t m_size{0};
public:
  bool isEmpty() const noexcept{
    return m_size == 0;
  }
  bool isFull() const noexcept{
    return m_size == Capacity;
  }
  void enqueue(T element){
    if(isFull()){
      throw std::overflow_error("Queue is Full");
    }
    m_data[m_rear] = std::move(element);
    std::cout << m_data[m_rear] << " enqueued to Queue.\n";
    m_rear = (m_rear + 1) % Capacity;
    ++m_size;
  }
  T dequeue(){
    if(isEmpty()){
      throw std::underflow_error("Queue is Empty !");
    }
    T element = std::move(m_data[m_front]);
    std::cout << "Element " << element << " removed from Queue.\n";
    m_front = (m_front + 1) % Capacity;
    --m_size;
    return element;
  }
};

int main(){
  try
  {
    CircularQueue<int,5> que;

    que.enqueue(11);
    que.enqueue(22);
    que.enqueue(33);
    que.enqueue(44);
    que.enqueue(55);

    que.dequeue();
    que.dequeue();

    que.enqueue(66);
    que.enqueue(77);

    que.enqueue(88); // this will hit queue overflow
  }
  catch(const std::exception& e)
  {
    std::cerr <<"Error: " << e.what() << '\n';
  }
  return 0;
}

File: Lab/Lab-3/Queue/Python/CircularQueue.py
class CircularQueue:
  def __init__(self,capacity:int):
    self.capacity = capacity
    self.data = [None] * capacity
    self.front = 0
    self.rear = 0
    self.size = 0


  def is_empty(self)->bool:
    return self.size == 0
  
  def is_full(self)->bool:
    return self.size == self.capacity
    
  def enqueue(self,element: int)->None:
    if(self.is_full()):
      raise OverflowError("Queue is Full !")
    
    self.data[self.rear] = element
    self.rear = (self.rear + 1) % self.capacity

    self.size += 1
    print(f"Enqueued element : {element}")

  def dequeue(self)->int:
    if(self.is_empty()):
      raise IndexError("Queue is Empty")
    
    element = self.data[self.front]
    self.front = (self.front + 1) % self.capacity
    self.size -= 1
    print(f"Dequeued Element : {element}")
    return element

try:
    q = CircularQueue(5)
    q.enqueue(11)
    q.enqueue(22)
    q.enqueue(33)
    q.enqueue(44)
    q.enqueue(55)
    
    # This will trigger the OverflowError
    q.enqueue(66) 

except OverflowError as e:
    print(f"Caught expected error: {e}")

try:
    q.dequeue()
    q.enqueue(66)
except IndexError as e:
    print(f"Error: {e}")